docs: ops tracing how to (#1853)
canonical/operator: docs/howto/tracing.md

(how-to-tracing)=
# How to trace charm code
## Introduction

This document describes how to trace your charm code and send the trace data to the [Canonical Observability Stack](https://documentation.ubuntu.com/observability/).

### Why trace your charms

> Observability transforms a Juju deployment from a black box to a real-time system.

Reviewer: I'm not sure the how-to needs to answer "why trace your charm" in general. But if we do include such a section, I don't love this sentence. Juju's real-time-ness is unrelated to whether it's observed or not. I don't think it's entirely fair to say that it's a black box without observability, either.

I think if we do include this section we should have a few concrete examples of problems that we're trying to solve (and ideally also some hints at what we are not trying to solve, like profiling).

However, it feels like the explanation is already working at answering this question, so we don't need it here too.

Trace data is structured and contextual; {ref}`this helps users <tracing-users>` understand the application's behaviour at different points in your charm's lifecycle.

### What is traced

The responsibility to instrument the Python code is divided between Ops, charm libraries and charms.

#### What Ops and charm libraries provide

Ops contains its own instrumentation, as well as unit and integration tests for it.

> - Ops creates the root span and a separate span when calling each observer, and these can be relied on in your tests.

Reviewer: I'm not sure what "relied on" means here. Do you mean that the spans are available for inspection/assertion in the tests? If so, I think a different phrase would be clearer. ", and these are collected during tests" or something like that?

Author: rewritten.

- Ops creates spans for every Juju hook tool invocation and every Pebble operation, so you typically don't have to.
- The {py:class}`ops.tracing.Tracing <ops_tracing.Tracing>` first-party charm library validates its arguments during object construction, and would trip on misconfiguration in any Scenario or integration test, so that charms don't have to test object instantiation explicitly.
- Buffering and export logic is already tested as part of Ops, and charms need not write integration
tests to cover these.

Charm libraries should instrument their logic at coarse granularity, and add attributes or events with logical information:
- the library object `__init__`, unless it's trivial, as that helps the charm authors
- spawned processes, that is `subprocess.*` calls with enough detail on what is being run, excluding sensitive data like the contents of keys or passwords
- outbound HTTP or RPC requests, as well as requests to own workload or peers, that is `requests.*` or `httpx.*` calls

> - important operations or operations that are intended to take time, like `bcrypt.hashpw()` or `ec.generate_private_key()`

Author: yes, definitely.

### What your charm needs to do

Charm authors should note that relatively little, if anything, needs to be instrumented in the charm code:
- workload decisions, such as failing over from primary to replica

> - external or long processes, like `subprocess.*`, `requests.*` or call to a Python module specific to your workload

Reviewer: ```suggestion
- external or long processes, like `subprocess.*`, `requests.*` or calls to a Python module specific to your workload
```

- points where important values can be exposed as attributes
- additional context for other instrumentation, like doing A for remote app B in a provider charm

> In a simple case, all the charm author needs to do is to instantiate {py:class}`ops.tracing.Tracing <ops_tracing.Tracing>` in the charm's `__init__` and provide it with relation names.

Author: Good idea to move a chunk of this to explanation; done!

I've attempted to rewrite the beginning/summary. It's shorter, though maybe doesn't look as great? I dunno.

## Add tracing to an existing charm

<!--
> Hints:
Updating an Existing Charm
Charm age and versions of library dependencies:
New enough Ops
Scenario tests
A word about integration tests

> -->

Reviewer: Same comment as the other commented bits :)

### To enable basic tracing

- In `pyproject.toml` or `requirements.txt`, add `ops[tracing]` as a dependency

> - In `charmcraft.yaml`, declare the relations for tracing and (optionally) certificate_transfer interfaces, for example:

Author: I'll leave this out for now.
It's a concern that charmers solve by blueprints (?) or some spec how their charm should be deployed.

(trace-your-charm)=
# How to trace your charm

This document describes how to trace your charm code and send the trace data to the [Canonical Observability Stack](https://documentation.ubuntu.com/observability/).

> Observability transforms a Juju deployment from a black box to a real-time system: trace data is structured and contextual; [this helps users](#tracing-users) understand the application's behaviour at different points in your charm's lifecycle.
> 
> The responsibility to instrument the Python code is divided between [Ops, charm libraries, and charms](#tracing-division-of-responsibilities).

Reviewer 2: A couple of minor suggestions for readability. I suggest replacing the inline links by a single "see also" link, to reduce people feeling like there's a lot of stuff they need to go and read in advance.

```suggestion
Observability transforms a Juju deployment from a black box to a real-time system. Trace data is structured and contextual, which helps users understand the application's behaviour at different points in your charm's lifecycle.

The responsibility to instrument the Python code is divided between Ops, charm libraries, and charms.

> See also: [](#tracing)
```

- Ops calls that inspect and update Juju (also called "hook tools")
- Pebble API access by the charm code

> ### Try it out

Author: Fair, done.

juju integrate my-charm tempo
```

### Fixme more

> ## Replace the charm_tracing library

Reviewer 2: Should this section be nested inside "Add tracing to an existing charm" instead? (ie one more `#`)

Author: Thanks, will work on it.

            ...
```

Refer to the {ref}`ops_tracing` reference for the canonical usage example, configuration
options, and API details.

> ### Replace the charm_tracing library

Reviewer: My feeling is that people would search for "migrating" (or "migrate", and hopefully the search has some stemming). Could we stick with that word?

of the decorated charm class. `ops[tracing]` doesn't do that, and you are expected
to create custom spans and events using the OpenTelemetry API where that makes sense.

<!--
## Trace a charm library

> Hints:
OTEL API usage.
A (dummy) charm to test the charm lib.

- At the top of your charm library, `import opentelemetry.trace`.
- After the imports in your charm library, create the tracer object as
  `tracer = opentelemetry.trace.get_tracer(name)` where the name could be your
  charm library name, or Python module `__name__`.
- See the [Custom spans and events](custom-spans-and-events) section above to
  create OpenTelemetry spans and events in the key places in your charm library.
-->

<!--
## Include tracing in a new charm
-->

<!--
## View the trace data

> Hints:
How to install COS (link), relate your app to COS, where to point browser,
screenshot of a trace in Grafana, some pointer about what's on the screenshot
and how manual instrumentation (above) connects to that.

> -->

Reviewer: Why is this here?

## Test the feature

A note that ops and charm libraries are already covered, maybe there's
nothing left for the charm to do.

Relation setup is already validated by the fact that your charm doesn't crash in other tests.

If you've added custom instrumentation, it is because something important is recorded.

> It's easier to cover this with unit tests.

Author: A note... was indeed a note to self.

xxx is already done... remains. I dunno if that's a good intro or not.

> It's easier to cover this with unit tests.

Reviewer 2: My feeling is that no intro is needed in this case

### Write unit tests

> The {py:attr}`trace_data <ops.testing.Context.trace_data>` attribute
> of the {py:class}`ops.testing.Context` class of the testing framework 
> contains the list of finished spans.

Reviewer: We generally write this like:

```suggestion
The [](ops.testing.Context.trace_data) attribute
of the [](ops.testing.Context) class of the testing framework 
contains the list of finished spans.
```

If `{py:}` is better for reason, I'm not opposed to switching to that, but I think it is nice if we're reasonably consistent across our docs.

Author: move to short syntax.

The following example demonstrates how to get the root span, created by Ops itself.

```py
ctx = Context(YourCharm)
ctx.run(ctx.on.start(), State())
main_span = next(s for s in ctx.trace_data if s.name == 'ops.main')
```

> The spans are OpenTelemetry objects in memory, allowing more focused examination, here a check that span A is a parent of span B.

Reviewer: ```suggestion
The spans are OpenTelemetry objects in memory, allowing more focused examination; here is a check that span A is a parent of span B.
```

```py
span_a = ...
span_b = ...
assert span_a.context is span_b.parent
```

Or that span A is an ancestor of span C, which allows you to validate that an important logical thing C was done during some well-known process A: an Ops event representing the Juju event or a custom event, or some manually instrumented function.

```py
spans_by_id = {s.context.span_id: s for s in ctx.trace_data}

def ancestors(span: ReadableSpan) -> Generator[ReadableSpan]:
    while span.parent:
        span = spans_by_id[span.parent.span_id]
        yield span

assert span_a in list(ancestors(span_c))
```

You can disambiguate spans using their {py:class}`instrumentation_scope <opentelemetry.sdk.util.instrumentation.InstrumentationScope>` property.

```py
# Spans from Ops
ops_span.instrumentation_scope.name == "ops"
ops_span.name == ...

# tracer = opentelemetry.trace.get_tracer("my-charm")
my_span.instrumentation_scope.name == "my-charm"
my_span.name == ...
```

<!--
### Write integration tests

> Hints:
Against the tracing-integration-tester charm (example)
Against Tempo
Links to COS documentation
> Against grafana-agent -- don't use that any more

> -->

Reviewer: If this is to be done in a later PR, then I'd rather we tracked that in a ticket than in a HTML comment.

> ## Lower-level API

Author: A separate PR, after the other (previously commented out) sections are done.
